packages/flutter_tools/lib/src/runner/flutter_command.dart

```diff
@@ -1076,9 +1089,23 @@ abstract class FlutterCommand extends Command<void> {
   Future<void> run() {
     final DateTime startTime = globals.systemClock.now();
 
+    FileSystem fileSystemOverride;
+
+    if (fileSystemScheme != null && fileSystemRoots?.isNotEmpty == true) {
+      fileSystemOverride = MultiRootFileSystem(
+        delegate: globals.fs,
+        scheme: fileSystemScheme,
+        roots: fileSystemRoots,
+      );
+    }
+
     return context.run<void>(
       name: 'command',
-      overrides: <Type, Generator>{FlutterCommand: () => this},
+      overrides: <Type, Generator>{
```

Comment:
I think this injection might be too late, since the package file path is parsed by the tool in the flutter command. What about hoisting the file system options into https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/executable.dart#L62 . Maybe we could create some kind of arg parser that ignored non-matching options?

Reply:
I would personally avoid adding more args parsing logic into that file. Any changes to that part, we need to manually apply internally, and parsing args outside just seems a bit hacky to me.

> the package file path is parsed by the tool in the flutter command

Which part exactly were you referring to? The only reference to any paths I can find in this file is in `verifyThenRunCommand` and that happens after this.

On an unrelated note, there are a few other caveats that I think is noteworthy. E.g. we can't call `fs.path.fromUri(uri)` or `uri.toFilePath()` since it would be a runtime error if a scheme URI is being used. I've personally tested `flutter run` a sample app with scheme URI both externally and internally, and I don't think anyone is going to use this externally, and we will have tests that cover our use cases internally so I think this should be good. These usages are usually easy to work around anyway.

Reply:
I don't know that we use fromUri anywhere but we may convert to file path in a few places. On the arg parsing - definitely, but long term I don't want to add more dependencies to the context injection system.

Not to mention injecting something again like the filesystem means that some objects may be created with a different file system based on when they were first accessed.

Backing up a bit - maybe the multiroot filesystem should allow adding a root/scheme after construction, and could default to a "regular" delegate file system if they weren't present. That would allow you to unconditionally set up the multiroot system and then configure it as soon as the args are parsed.